node/subsystem-util/src/database.rs

```diff
@@ -0,0 +1,335 @@
+// Copyright 2021-2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Database trait for polkadot db.
+
+pub use kvdb::{DBTransaction, DBValue, KeyValueDB};
+
+/// Database trait with ordered key capacity.
+pub trait Database: KeyValueDB {
+	/// Check if column allows content iteration
+	/// and removal by prefix.
+	fn is_indexed_column(&self, col: u32) -> bool;
+}
+
+/// Implementation for database supporting `KeyValueDB` already.
+pub mod kvdb_impl {
+	use super::{DBTransaction, DBValue, Database, KeyValueDB};
+	use kvdb::{DBOp, IoStats, IoStatsKind};
+	use parity_util_mem::{MallocSizeOf, MallocSizeOfOps};
+	use std::{collections::BTreeSet, io::Result};
+
+	/// Adapter implementing subsystem database
+	/// for `KeyValueDB`.
+	#[derive(Clone)]
+	pub struct DbAdapter<D> {
+		db: D,
+		allowed_iter: BTreeSet<u32>,
+	}
+
+	impl<D: KeyValueDB> DbAdapter<D> {
+		/// Instantiate new subsystem database, with
+		/// definition of collections allowing iteration.
+		pub fn new(db: D, allowed_iter: &[u32]) -> Self {
+			DbAdapter { db, allowed_iter: allowed_iter.iter().cloned().collect() }
+		}
+
+		fn filter_iter(&self, col: u32) {
+			debug_assert!(
```

Comment:
Actually `debug_assert` is not that usefull since CI probably run in release mode.

Reply:
It runs with `--profile testnet` which enables debug assertions, but previously they were also enabled via RUSTFLAGS